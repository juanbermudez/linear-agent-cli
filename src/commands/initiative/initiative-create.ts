import { Command } from "@cliffy/command"
import { Input, Select } from "@cliffy/prompt"
import {
  createInitiative,
  listInitiativeStatuses,
  listUsers,
} from "../../utils/linear.ts"
import {
  error as errorColor,
  success as successColor,
} from "../../utils/styling.ts"

interface CreateOptions {
  name?: string
  description?: string
  content?: string
  status?: string
  owner?: string
  targetDate?: string
  json?: boolean
  format?: string
  noInteractive?: boolean
}

export const createCommand = new Command()
  .name("create")
  .description("Create a new initiative")
  .option("-n, --name <name:string>", "Initiative name")
  .option("-d, --description <description:string>", "Initiative description")
  .option(
    "-c, --content <content:string>",
    "Initiative content/body (markdown)",
  )
  .option("-s, --status <status:string>", "Status ID or name")
  .option("-o, --owner <owner:string>", "Owner user ID or display name")
  .option("-t, --target-date <date:string>", "Target date (YYYY-MM-DD)")
  .option("--no-interactive", "Disable interactive mode")
  .option("-j, --json", "Output result as JSON")
  .option("--format <format:string>", "Output format: text|json")
  .action(async (options: CreateOptions) => {
    const useJson = options.json || options.format === "json"
    const interactive = !options.name && !useJson && !options.noInteractive &&
      Deno.stdout.isTerminal()

    let name: string
    let description: string | undefined
    let content: string | undefined
    let statusId: string | undefined
    let ownerId: string | undefined
    let targetDate: string | undefined

    if (interactive) {
      name = await Input.prompt({
        message: "Initiative name",
        validate: (value) => value.trim() !== "" || "Name is required",
      })

      description = await Input.prompt({
        message: "Description (optional)",
        default: "",
      })
      if (description === "") description = undefined

      // Status selection
      const { Spinner } = await import("@std/cli/unstable-spinner")
      const statusSpinner = new Spinner({ message: "Loading statuses..." })
      statusSpinner.start()
      const statuses = await listInitiativeStatuses()
      statusSpinner.stop()

      if (statuses.length > 0) {
        const statusOptions = statuses.map((s) => ({
          name: `${s.name} (${s.type})`,
          value: s.id,
        }))
        statusOptions.unshift({ name: "None", value: "" })

        const selectedStatus = await Select.prompt({
          message: "Initiative status (optional)",
          options: statusOptions,
          search: true,
        })
        statusId = selectedStatus || undefined
      }

      // Owner selection
      const ownerSpinner = new Spinner({ message: "Loading users..." })
      ownerSpinner.start()
      const users = await listUsers()
      ownerSpinner.stop()

      const ownerOptions = users.map((u) => ({
        name: u.displayName,
        value: u.id,
      }))
      ownerOptions.unshift({ name: "None", value: "" })

      const selectedOwner = await Select.prompt({
        message: "Initiative owner (optional)",
        options: ownerOptions,
        search: true,
      })
      ownerId = selectedOwner || undefined

      // Target date
      targetDate = await Input.prompt({
        message: "Target date (YYYY-MM-DD, optional)",
        default: "",
        validate: (value) => {
          if (!value) return true
          if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) {
            return "Invalid date format. Use YYYY-MM-DD"
          }
          return true
        },
      })
      if (targetDate === "") targetDate = undefined
    } else {
      if (!options.name) {
        const errorMsg = "Name is required. Use --name or run without flags"
        if (useJson) {
          console.error(
            JSON.stringify(
              {
                success: false,
                error: {
                  code: "MISSING_REQUIRED_FIELD",
                  message: errorMsg,
                },
              },
              null,
              2,
            ),
          )
        } else {
          console.error(errorColor(`Error: ${errorMsg}`))
        }
        Deno.exit(1)
      }

      name = options.name
      description = options.description
      content = options.content
      targetDate = options.targetDate

      // Validate date format if provided
      if (targetDate && !/^\d{4}-\d{2}-\d{2}$/.test(targetDate)) {
        const errorMsg = "Invalid date format. Use YYYY-MM-DD"
        if (useJson) {
          console.error(
            JSON.stringify(
              {
                success: false,
                error: { code: "INVALID_VALUE", message: errorMsg },
              },
              null,
              2,
            ),
          )
        } else {
          console.error(errorColor(`Error: ${errorMsg}`))
        }
        Deno.exit(1)
      }

      // Resolve status if provided
      if (options.status) {
        if (
          /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i
            .test(options.status)
        ) {
          statusId = options.status
        } else {
          const statuses = await listInitiativeStatuses()
          const status = statuses.find((s) =>
            s.name.toLowerCase() === options.status!.toLowerCase()
          )
          if (status) {
            statusId = status.id
          } else {
            const errorMsg = `Status '${options.status}' not found`
            if (useJson) {
              console.error(
                JSON.stringify(
                  {
                    success: false,
                    error: { code: "NOT_FOUND", message: errorMsg },
                  },
                  null,
                  2,
                ),
              )
            } else {
              console.error(errorColor(`Error: ${errorMsg}`))
            }
            Deno.exit(1)
          }
        }
      }

      // Resolve owner if provided
      if (options.owner) {
        if (
          /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i
            .test(options.owner)
        ) {
          ownerId = options.owner
        } else {
          const users = await listUsers()
          const user = users.find((u) =>
            u.displayName.toLowerCase() === options.owner!.toLowerCase()
          )
          if (user) {
            ownerId = user.id
          } else {
            const errorMsg = `User '${options.owner}' not found`
            if (useJson) {
              console.error(
                JSON.stringify(
                  {
                    success: false,
                    error: { code: "NOT_FOUND", message: errorMsg },
                  },
                  null,
                  2,
                ),
              )
            } else {
              console.error(errorColor(`Error: ${errorMsg}`))
            }
            Deno.exit(1)
          }
        }
      }
    }

    const { Spinner } = await import("@std/cli/unstable-spinner")
    const showSpinner = !useJson && Deno.stdout.isTerminal()
    const spinner = showSpinner
      ? new Spinner({ message: "Creating initiative..." })
      : null
    spinner?.start()

    try {
      const initiative = await createInitiative({
        name,
        description,
        content,
        statusId,
        ownerId,
        targetDate,
      })

      spinner?.stop()

      if (useJson) {
        console.log(
          JSON.stringify(
            {
              success: true,
              operation: "create",
              initiative: {
                id: initiative.id,
                name: initiative.name,
                slugId: initiative.slugId,
                url: initiative.url,
                status: initiative.status
                  ? { id: initiative.status.id, name: initiative.status.name }
                  : null,
                owner: initiative.owner
                  ? {
                    id: initiative.owner.id,
                    name: initiative.owner.displayName,
                  }
                  : null,
                targetDate: initiative.targetDate,
              },
            },
            null,
            2,
          ),
        )
      } else {
        console.log(successColor(`âœ“ Created initiative ${initiative.slugId}`))
        console.log(initiative.url)
      }
    } catch (err) {
      spinner?.stop()
      if (useJson) {
        console.error(
          JSON.stringify(
            {
              success: false,
              error: { code: "API_ERROR", message: (err as Error).message },
            },
            null,
            2,
          ),
        )
      } else {
        console.error(errorColor(`Error: ${(err as Error).message}`))
      }
      Deno.exit(1)
    }
  })
