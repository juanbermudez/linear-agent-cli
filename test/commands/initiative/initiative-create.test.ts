import { snapshotTest } from "@cliffy/testing"
import { createCommand } from "../../../src/commands/initiative/initiative-create.ts"
import {
  commonDenoArgs,
  setupMockLinearServer,
} from "../../utils/test-helpers.ts"

// Test help output
await snapshotTest({
  name: "Initiative Create Command - Help Text",
  meta: import.meta,
  colors: false,
  args: ["--help"],
  denoArgs: commonDenoArgs,
  async fn() {
    await createCommand.parse()
  },
})

// Test creating an initiative with flags (happy path)
await snapshotTest({
  name: "Initiative Create Command - Happy Path",
  meta: import.meta,
  colors: false,
  args: [
    "--name",
    "Q1 2025 Platform Goals",
    "--description",
    "Focus on platform stability and performance",
    "--owner",
    "alice@company.com",
    "--status",
    "planned",
    "--json",
    "--no-color",
  ],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      // Mock lookupUserId for owner
      {
        queryName: "LookupUser",
        variables: { input: "alice@company.com" },
        response: {
          data: {
            users: {
              nodes: [{
                id: "user-alice-123",
                email: "alice@company.com",
                displayName: "Alice Smith",
                name: "alice",
              }],
            },
          },
        },
      },
      // Mock listInitiativeStatuses
      {
        queryName: "ListInitiativeStatuses",
        response: {
          data: {
            workflowStates: {
              nodes: [
                { id: "status-planned", name: "Planned", type: "planned" },
                { id: "status-active", name: "Active", type: "started" },
                {
                  id: "status-completed",
                  name: "Completed",
                  type: "completed",
                },
              ],
            },
          },
        },
      },
      // Mock createInitiative mutation
      {
        queryName: "CreateInitiative",
        response: {
          data: {
            initiativeCreate: {
              success: true,
              initiative: {
                id: "init-123",
                name: "Q1 2025 Platform Goals",
                slugId: "q1-2025-platform-goals",
                url:
                  "https://linear.app/test/initiative/q1-2025-platform-goals",
                description: "Focus on platform stability and performance",
                status: {
                  id: "status-planned",
                  name: "Planned",
                  type: "planned",
                },
                owner: {
                  id: "user-alice-123",
                  name: "alice",
                  displayName: "Alice Smith",
                },
                createdAt: "2024-01-15T10:00:00Z",
                updatedAt: "2024-01-15T10:00:00Z",
              },
            },
          },
        },
      },
    ])

    try {
      await createCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test creating an initiative with minimal flags
await snapshotTest({
  name: "Initiative Create Command - Minimal Flags",
  meta: import.meta,
  colors: false,
  args: [
    "--name",
    "Q2 Goals",
    "--json",
    "--no-color",
  ],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "CreateInitiative",
        response: {
          data: {
            initiativeCreate: {
              success: true,
              initiative: {
                id: "init-456",
                name: "Q2 Goals",
                slugId: "q2-goals",
                url: "https://linear.app/test/initiative/q2-goals",
                description: null,
                status: null,
                owner: null,
                createdAt: "2024-01-15T10:00:00Z",
                updatedAt: "2024-01-15T10:00:00Z",
              },
            },
          },
        },
      },
    ])

    try {
      await createCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test creating initiative with target date
await snapshotTest({
  name: "Initiative Create Command - With Target Date",
  meta: import.meta,
  colors: false,
  args: [
    "--name",
    "Q1 Goals",
    "--target-date",
    "2025-03-31",
    "--json",
    "--no-color",
  ],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "CreateInitiative",
        response: {
          data: {
            initiativeCreate: {
              success: true,
              initiative: {
                id: "init-789",
                name: "Q1 Goals",
                slugId: "q1-goals",
                url: "https://linear.app/test/initiative/q1-goals",
                description: null,
                status: null,
                owner: null,
                targetDate: "2025-03-31",
                createdAt: "2024-01-15T10:00:00Z",
                updatedAt: "2024-01-15T10:00:00Z",
              },
            },
          },
        },
      },
    ])

    try {
      await createCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test non-JSON output format
await snapshotTest({
  name: "Initiative Create Command - Text Output",
  meta: import.meta,
  colors: false,
  args: [
    "--name",
    "Platform Initiative",
    "--no-color",
  ],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "CreateInitiative",
        response: {
          data: {
            initiativeCreate: {
              success: true,
              initiative: {
                id: "init-text",
                name: "Platform Initiative",
                slugId: "platform-initiative",
                url: "https://linear.app/test/initiative/platform-initiative",
                description: null,
                status: null,
                owner: null,
                createdAt: "2024-01-15T10:00:00Z",
                updatedAt: "2024-01-15T10:00:00Z",
              },
            },
          },
        },
      },
    ])

    try {
      await createCommand.parse()
    } finally {
      await cleanup()
    }
  },
})
