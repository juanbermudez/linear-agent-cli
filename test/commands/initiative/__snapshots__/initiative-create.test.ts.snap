export const snapshot = {};

snapshot[`Initiative Create Command - Help Text 1`] = `
stdout:
"
Usage: create

Description:

  Create a new initiative

Options:

  -h, --help                        - Show this help.                     
  -n, --name         <name>         - Initiative name                     
  -d, --description  <description>  - Initiative description              
  -c, --content      <content>      - Initiative content/body (markdown)  
  -s, --status       <status>       - Status ID or name                   
  -o, --owner        <owner>        - Owner user ID or display name       
  -t, --target-date  <date>         - Target date (YYYY-MM-DD)            
  --no-interactive                  - Disable interactive mode            
  --no-color                        - Disable colored output              
  -j, --json                        - Output result as JSON               
  --format           <format>       - Output format: text|json            

"
stderr:
""
`;

snapshot[`Initiative Create Command - Happy Path 1`] = `
stdout:
'{
  "success": true,
  "operation": "create",
  "initiative": {
    "id": "init-123",
    "name": "Q1 2025 Platform Goals",
    "slugId": "q1-2025-platform-goals",
    "url": "https://linear.app/test/initiative/q1-2025-platform-goals",
    "status": {
      "id": "status-planned",
      "name": "Planned",
      "type": "planned"
    },
    "owner": {
      "id": "user-alice-123",
      "name": "Alice Smith"
    }
  }
}
'
stderr:
""
`;

snapshot[`Initiative Create Command - Minimal Flags 1`] = `
stdout:
'{
  "success": true,
  "operation": "create",
  "initiative": {
    "id": "init-456",
    "name": "Q2 Goals",
    "slugId": "q2-goals",
    "url": "https://linear.app/test/initiative/q2-goals",
    "status": null,
    "owner": null
  }
}
'
stderr:
""
`;

snapshot[`Initiative Create Command - With Target Date 1`] = `
stdout:
'{
  "success": true,
  "operation": "create",
  "initiative": {
    "id": "init-789",
    "name": "Q1 Goals",
    "slugId": "q1-goals",
    "url": "https://linear.app/test/initiative/q1-goals",
    "status": null,
    "owner": null,
    "targetDate": "2025-03-31"
  }
}
'
stderr:
""
`;

snapshot[`Initiative Create Command - Text Output 1`] = `
stdout:
"âœ“ Created initiative platform-initiative
https://linear.app/test/initiative/platform-initiative
"
stderr:
""
`;
